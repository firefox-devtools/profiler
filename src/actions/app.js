/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { oneLine } from 'common-tags';
import {
  getSelectedTab,
  getDataSource,
  getLocalTrackOrderByPid,
} from 'firefox-profiler/selectors/url-state';
import {
  getTrackThreadHeights,
  getIsEventDelayTracksEnabled,
  getIsExperimentalCPUGraphsEnabled,
  getIsExperimentalProcessCPUTracksEnabled,
} from 'firefox-profiler/selectors/app';
import {
  getLocalTracksByPid,
  getThreads,
  getCounters,
} from 'firefox-profiler/selectors/profile';
import { sendAnalytics } from 'firefox-profiler/utils/analytics';
import {
  stateFromLocation,
  withHistoryReplaceStateSync,
} from 'firefox-profiler/app-logic/url-handling';
import { finalizeProfileView } from './receive-profile';
import { fatalError } from './errors';
import {
  addEventDelayTracksForThreads,
  initializeLocalTrackOrderByPid,
  addProcessCPUTracksForProcess,
} from 'firefox-profiler/profile-logic/tracks';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import {
  getIsCPUUtilizationProvided,
  getAreThereAnyProcessCPUCounters,
} from 'firefox-profiler/selectors/cpu';

import type {
  Profile,
  ThreadsKey,
  CssPixels,
  Action,
  ThunkAction,
  UrlState,
  UploadedProfileInformation,
  IndexIntoCategoryList,
} from 'firefox-profiler/types';
import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';
import type {
  BrowserConnection,
  BrowserConnectionStatus,
} from 'firefox-profiler/app-logic/browser-connection';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function startFetchingProfiles(): Action {
  return { type: 'START_FETCHING_PROFILES' };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when we start setting up the initial url state.
 * It takes the location and profile data, converts the location into url
 * state and then dispatches relevant actions to finalize the view.
 * `profile` parameter can be null when the data source can't provide the profile
 * and the url upgrader step is not needed (e.g. 'from-browser').
 */
export function setupInitialUrlState(
  location: Location,
  profile: Profile | null,
  browserConnection: BrowserConnection | null
): ThunkAction<void> {
  return (dispatch) => {
    let urlState;
    try {
      urlState = stateFromLocation(location, profile);
    } catch (e) {
      if (e.name === 'UrlUpgradeError') {
        // The error is an URL upgrade error, let's fire a fatal error.
        // If there's a service worker update, the class `ServiceWorkerManager`
        // will automatically reload in case the new code knows how to handle
        // this URL version.
        dispatch(fatalError(e));
        return;
      }
      // The location could not be parsed, show a 404 instead.
      console.error(e);
      dispatch(show404(location.pathname + location.search));
      return;
    }

    // Validate the initial URL state. We can't refresh on a from-file URL.
    if (urlState.dataSource === 'from-file') {
      urlState = null;
    }

    // Normally having multiple dispatches is an anti pattern, but here it's
    // necessary because we are doing different things inside those actions and
    // they can't be merged because we are also calling those seperately on
    // other parts of the code.
    // The first dispatch here updates the url state, then changes state as the url
    // setup is done, and lastly finalizes the profile view since everything is set up now.
    // All of this is done while the history is replaced, as this is part of the initial
    // load process.
    withHistoryReplaceStateSync(() => {
      dispatch(updateUrlState(urlState));
      dispatch(finalizeProfileView(browserConnection));
      dispatch(urlSetupDone());
    });
  };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export function reportTrackThreadHeight(
  threadsKey: ThreadsKey,
  height: CssPixels
): ThunkAction<void> {
  return (dispatch, getState) => {
    const trackThreadHeights = getTrackThreadHeights(getState());
    const previousHeight = trackThreadHeights[threadsKey];
    if (previousHeight !== height) {
      // Guard against unnecessary dispatches. This could happen frequently.
      dispatch({
        type: 'UPDATE_TRACK_THREAD_HEIGHT',
        height,
        threadsKey,
      });
    }
  };
}

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}

/**
 * Called when a user has started dragging a file. Used for loading
 * profiles with the drag and drop component.
 */
export function startDragging(): Action {
  return { type: 'START_DRAGGING' };
}

/**
 * Called when a user has stopped dragging a file.
 */
export function stopDragging(): Action {
  return { type: 'STOP_DRAGGING' };
}

/**
 * Called when a custom drag and drop overlay is mounted. This lets
 * the app know that we shouldn't create a default overlay.
 */
export function registerDragAndDropOverlay(): Action {
  return { type: 'REGISTER_DRAG_AND_DROP_OVERLAY' };
}

/**
 * Called when a custom drag and drop overlay is unmounted.
 */
export function unregisterDragAndDropOverlay(): Action {
  return { type: 'UNREGISTER_DRAG_AND_DROP_OVERLAY' };
}

/*
 * This action enables the event delay tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableEventDelayTracks()`
 */
export function enableEventDelayTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsEventDelayTracksEnabled(getState())) {
      console.error(
        'Tried to enable the event delay tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      selectedThreadSelectors.getSamplesTable(getState()).eventDelay ===
      undefined
    ) {
      // Return early if the profile doesn't have eventDelay values.
      console.error(oneLine`
        Tried to enable the event delay tracks, but this profile does
        not have eventDelay values. It is likely an older profile.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addEventDelayTracksForThreads(
      getThreads(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );
    dispatch({
      type: 'ENABLE_EVENT_DELAY_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/*
 * This action enables the CPU graph tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableCPUGraphs()`
 */
export function enableExperimentalCPUGraphs(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalCPUGraphsEnabled(getState())) {
      console.error(
        'Tried to enable the CPU graph tracks, but they are already enabled.'
      );
      return false;
    }

    if (!getIsCPUUtilizationProvided(getState())) {
      // Return early if the profile doesn't have threadCPUDelta values.
      console.error(oneLine`
        Tried to enable the CPU graph tracks, but this profile does
        not have threadCPUDelta values. It is likely an older profile.
      `);
      return false;
    }

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_CPU_GRAPHS',
    });

    return true;
  };
}

/*
 * This action enables the process CPU tracks. They are hidden by default because
 * the front-end work is not done for this data yet.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableExperimentalProcessCPUTracks()`
 */
export function enableExperimentalProcessCPUTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalProcessCPUTracksEnabled(getState())) {
      console.error(
        'Tried to enable the process CPU tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      !getIsCPUUtilizationProvided(getState()) &&
      getAreThereAnyProcessCPUCounters(getState())
    ) {
      // Return early if the profile doesn't have threadCPUDelta values or
      // doesn't have any experimental process CPU counters.
      console.error(oneLine`
        Tried to enable the process CPU tracks, but this profile does
        not have threadCPUDelta values or process CPU threads.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addProcessCPUTracksForProcess(
      getCounters(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_PROCESS_CPU_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/**
 * This caches the profile data in the local state for synchronous access.
 */
export function setCurrentProfileUploadedInformation(
  uploadedProfileInformation: UploadedProfileInformation | null
): Action {
  return {
    type: 'SET_CURRENT_PROFILE_UPLOADED_INFORMATION',
    uploadedProfileInformation,
  };
}

export function profileRemotelyDeleted(): Action {
  // Ideally we should store the current profile data in a local indexeddb, and
  // set the URL to /local/<indexeddb-key>.
  return { type: 'PROFILE_REMOTELY_DELETED' };
}

export function updateBrowserConnectionStatus(
  browserConnectionStatus: BrowserConnectionStatus
): Action {
  return { type: 'UPDATE_BROWSER_CONNECTION_STATUS', browserConnectionStatus };
}

export function toggleOpenCategoryInSidebar(
  kind: string,
  category: IndexIntoCategoryList
): Action {
  return {
    type: 'TOGGLE_SIDEBAR_OPEN_CATEGORY',
    kind,
    category,
  };
}
