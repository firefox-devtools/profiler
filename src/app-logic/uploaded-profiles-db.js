/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

// This file contains the code responsible for storing informations about
// published profiles.

import { openDB, deleteDB } from 'idb';
import { stripIndent } from 'common-tags';
import {
  stateFromLocation,
  urlFromState,
} from 'firefox-profiler/app-logic/url-handling';
import { ensureExists } from 'firefox-profiler/utils/flow';

import type { DB as Database } from 'idb';
import type { StartEndRange } from 'firefox-profiler/types';

// This type is closely tied to the IndexedDB operation. Indeed it represents
// the data we store and retrieve in the local DB. That's especially why it's
// defined in this file, close to the DB operations. Indeed we don't want that
// this type evolves without implementing a migration step for the stored data.
export type UploadedProfileInformation = {|
  +profileToken: string, // This is the primary key.
  +jwtToken: string | null,
  +publishedDate: Date, // This key is indexed as well, to provide automatic sorting.
  +name: string,
  +preset: string | null,
  +meta: {|
    // We're using some of the properties of the profile meta, but we're not
    // reusing the type ProfileMeta completely because we don't want to be
    // impacted from future changes to ProfileMeta.
    // Look at ProfileMeta definition to know more about these fields.
    +product: string,
    +abi?: string,
    +platform?:
      | 'Android'
      | 'Windows'
      | 'Macintosh'
      // X11 is used for historic reasons, but this value means that it is a Unix platform.
      | 'X11'
      | string,
    +toolkit?: string,
    +misc?: string,
    +oscpu?: string,
    // Older versions of Firefox for Linux had the 2 flavors gtk2/gtk3, and so
    // we could find the value "gtk3".
    +toolkit?: 'gtk' | 'gtk3' | 'windows' | 'cocoa' | 'android' | string,
    +updateChannel?:
      | 'default' // Local builds
      | 'nightly'
      | 'nightly-try' // Nightly try builds for QA
      | 'aurora' // Developer Edition channel
      | 'beta'
      | 'release'
      | 'esr' // Extended Support Release channel
      | string,
    +appBuildID?: string,
  |},
  // Storing the state as the path makes it easy to reuse our URL upgrade mechanism.
  +urlPath: string,
  +publishedRange: StartEndRange,
|};

// Exported for tests.
export const DATABASE_NAME = 'published-profiles-store';
export const OBJECTSTORE_NAME = 'published-profiles';
export const DATABASE_VERSION = 3;

async function reallyOpen(): Promise<Database> {
  const db = await openDB(DATABASE_NAME, DATABASE_VERSION, {
    upgrade(db, oldVersion, newVersion, transaction) {
      // In the following switch block, we don't "break" for each case block
      // because we want to run all migration steps in sequence, starting with
      // the right step.
      /* eslint-disable no-fallthrough */
      switch (oldVersion) {
        case 0: {
          // Version 1: this is the first version of the DB.
          const store = db.createObjectStore(OBJECTSTORE_NAME, {
            keyPath: 'profileToken',
          });
          store.createIndex('originHostname', 'originHostname');
        }
        case 1: {
          // Version 2: we create a new index to allow retrieving the values
          // ordered by date.
          const store = ensureExists(transaction.store);
          store.createIndex('publishedDate', 'publishedDate');
        }
        case 2: {
          // Version 3: we remove the originHostname index that was used by the
          // active tab view since it's been removed.
          const store = ensureExists(transaction.store);
          store.deleteIndex('originHostname');
        }
        default:
        // Nothing more here.
      }
      /* eslint-enable */
    },
  });

  return db;
}

async function open(): Promise<Database> {
  if (!window.indexedDB) {
    throw new Error('Could not find indexedDB on the window object.');
  }

  let db;
  try {
    db = await reallyOpen();
  } catch (e) {
    if (e.name === 'VersionError') {
      // This error fires if the database already exists, and the existing
      // database has a higher version than what we requested. So either
      // this version of profiler.firefox.com is outdated, or somebody briefly tried
      // to change this database format (and increased the version number)
      // and then downgraded to a version of profiler.firefox.com without those
      // changes.
      // Let's explain that in an error, that will be output to the console by
      // the caller.
      (window: any).deleteDB = () => deleteDB(DATABASE_NAME);
      throw new Error(stripIndent`
        We tried to open an existing published profiles store database with a
        smaller version than the current one. We can't do that with IndexedDB.
        The only way to recover is to delete the database and create a new one,
        but we don't want to do this automatically.
        Until this is fixed we won't be able to store newly published profiles
        or retrieve previously published profiles.

        If you want to delete the database, you can do so by running 'deleteDB()'
        in the console, or using the developer tools.
      `);
    }
    throw e;
  }

  return db;
}

/**
 * This stores some profile data. The profileToken property is the primary key,
 * so this also updates any profile data already there with the same
 * profileToken information.
 */
export async function persistUploadedProfileInformationToDb(
  uploadedProfileInformation: UploadedProfileInformation
): Promise<void> {
  const db = await open();
  await db.put(OBJECTSTORE_NAME, uploadedProfileInformation);
}

/**
 * This returns the list of all the stored data.
 */
export async function listAllUploadedProfileInformationFromDb(): Promise<
  UploadedProfileInformation[],
> {
  const db = await open();
  return db.getAllFromIndex(OBJECTSTORE_NAME, 'publishedDate');
}

/**
 * This returns the profile data for a specific stored token, or undefined
 * otherwise.
 */
export async function retrieveUploadedProfileInformationFromDb(
  profileToken: string
): Promise<UploadedProfileInformation | null> {
  if (!profileToken) {
    // If this is the empty string, let's skip the lookup.
    return null;
  }

  const db = await open();
  const result = await db.get(OBJECTSTORE_NAME, profileToken);
  return result || null;
}

/**
 * This deletes the profile data stored with this token. This is a no-op if this
 * token isn't in the database.
 */
export async function deleteUploadedProfileInformationFromDb(
  profileToken: string
): Promise<void> {
  const db = await open();
  return db.delete(OBJECTSTORE_NAME, profileToken);
}

/**
 * This changes the profile name of a stored profile data. This is a no-op if
 * this token isn't in the database.
 */
export async function changeStoredProfileNameInDb(
  profileToken: string,
  profileName: string
): Promise<void> {
  const storedProfile =
    await retrieveUploadedProfileInformationFromDb(profileToken);
  if (storedProfile && storedProfile.name !== profileName) {
    // We need to update the name, but also the urlPath. For this we'll convert
    // the old one to a state, and convert it back to a url string, so that
    // there is less chance that we forget about this case if we update the
    // state object.

    // `stateFromLocation` waits for something that looks like a Location
    // object. We use the URL object for this, but it requires a full URL, even
    // if `stateFromLocation` doesn't need one.
    const oldState = stateFromLocation(
      new URL(storedProfile.urlPath, window.location.href)
    );
    const newUrlPath = urlFromState({ ...oldState, profileName });

    const newUploadedProfileInformation = {
      ...storedProfile,
      name: profileName,
      urlPath: newUrlPath,
    };
    await persistUploadedProfileInformationToDb(newUploadedProfileInformation);
  }
}
