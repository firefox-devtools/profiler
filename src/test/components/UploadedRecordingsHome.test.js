/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React from 'react';
import { Provider } from 'react-redux';

import { render } from 'firefox-profiler/test/fixtures/testing-library';
import { UploadedRecordingsHome } from 'firefox-profiler/components/app/UploadedRecordingsHome';
import { persistUploadedProfileInformationToDb } from 'firefox-profiler/app-logic/uploaded-profiles-db';
import { blankStore } from '../fixtures/stores';
import { mockDate } from '../fixtures/mocks/date';

import { autoMockIndexedDB } from 'firefox-profiler/test/fixtures/mocks/indexeddb';
autoMockIndexedDB();

describe('UploadedRecordingsHome', () => {
  function setup() {
    const store = blankStore();
    const renderResult = render(
      <Provider store={store}>
        <UploadedRecordingsHome />
      </Provider>
    );
    return renderResult;
  }

  it('matches a snapshot when there is no published profiles', async () => {
    const { container, findByText } = setup();
    await findByText(/No profile has been uploaded/);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches a snapshot when rendering published profiles', async () => {
    // Because the rendering is dependent on the timezone, all dates in this
    // test are using a timezone-dependent format so that the rendering is the
    // same in all timezones.
    mockDate('4 Jul 2020 15:00'); // Now is 4th of July, at 3pm local timezone.

    // 1. Add some profiles to the local indexeddb.
    await persistUploadedProfileInformationToDb({
      profileToken: '0123456789',
      jwtToken: 'FAKE_TOKEN',
      publishedDate: new Date('4 Jul 2020 14:00'), // "today" earlier
      name: '',
      preset: null,
      meta: {
        product: 'Fennec',
        // Not more meta information, to test that we can handle profiles that
        // don't have this information.
      },
      urlPath: '/',
      publishedRange: { start: 1000, end: 3000 },
    });

    await persistUploadedProfileInformationToDb({
      profileToken: 'ABCDEFGHI',
      jwtToken: null,
      publishedDate: new Date('3 Jul 2020 08:00'), // yesterday
      name: 'Layout profile',
      preset: 'web',
      meta: {
        product: 'Firefox',
        platform: 'X11',
        toolkit: 'gtk',
        oscpu: 'Linux x86_64',
        misc: 'rv:68.0',
      },
      urlPath: '/',
      publishedRange: { start: 1000, end: 1005 },
    });

    await persistUploadedProfileInformationToDb({
      profileToken: '123abc456',
      jwtToken: null,
      publishedDate: new Date('20 May 2018'), // ancient date
      name: '',
      preset: null,
      meta: {
        product: 'Firefox Preview',
        platform: 'Android 7.0',
        toolkit: 'android',
        misc: 'rv:80.0',
        oscpu: 'Linux armv7l',
      },
      urlPath:
        '/public/123abc456/calltree/?globalTrackOrder=0-1-2-3-4-5-6-7-8-9-10&hiddenGlobalTracks=1-2-3-4-5-6-7-8-9&localTrackOrderByPid=32027-1-2-0~12629-0~12149-0~11755-0~11945-0~12533-0~11798-0~11862-0~12083-0~12118-0~12298-0-1~&search=bla&thread=11&v=5',
      // This tests ranges that should show up as Âµs
      publishedRange: { start: 40000, end: 40000.1 },
    });

    await persistUploadedProfileInformationToDb({
      profileToken: 'WINDOWS',
      jwtToken: null,
      publishedDate: new Date('4 Jul 2020 13:00'),
      name: 'Another good profile',
      preset: null,
      meta: {
        product: 'Firefox',
        platform: 'Windows',
        toolkit: 'windows',
        misc: 'rv:77.0',
        oscpu: 'Windows NT 10.0; Win64; x64',
      },
      urlPath: '/public/WINDOWS/marker-chart/',
      publishedRange: { start: 2000, end: 40000 },
    });

    await persistUploadedProfileInformationToDb({
      profileToken: 'MACOSX',
      jwtToken: null,
      publishedDate: new Date('5 Jul 2020 11:00'), // This is the future!
      name: 'MacOS X profile',
      preset: null,
      meta: {
        product: 'Firefox',
        platform: 'Macintosh',
        toolkit: 'cocoa',
        misc: 'rv:62.0',
        oscpu: 'Intel Mac OS X 10.12',
      },
      urlPath: '/public/MACOSX/marker-chart/',
      publishedRange: { start: 2000, end: 40000 },
    });

    // 2. Render the component and test.
    const { container, findByText } = setup();
    await findByText(/macOS/);

    expect(container.firstChild).toMatchSnapshot();
  });
});
