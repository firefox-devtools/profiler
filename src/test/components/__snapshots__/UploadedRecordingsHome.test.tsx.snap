// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UploadedRecordingsHome matches a snapshot when rendering published profiles 1`] = `
<main
  class="uploadedRecordingsHome"
>
  <header>
    <h1
      class="appHeader"
    >
      <span
        class="appHeaderSlogan"
      >
        <a
          class="appHeaderLink"
          href="/"
        >
          ⁨Firefox Profiler⁩
        </a>
         — 
        <span
          class="appHeaderSubtext"
        >
          Web app for ⁨Firefox⁩ performance analysis
        </span>
      </span>
      <a
        class="appHeaderGithubIcon"
        href="https://github.com/firefox-devtools/profiler"
        rel="noopener noreferrer"
        target="_blank"
        title="Go to our Git repository (this opens in a new window)"
      >
        <svg
          aria-label="github"
          class="octicon octicon-mark-github"
          height="22"
          version="1.1"
          viewBox="0 0 16 16"
          width="22"
        >
          <path
            d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"
            fill-rule="evenodd"
          />
        </svg>
      </a>
    </h1>
  </header>
  <h2
    class="photon-title-30"
  >
    Uploaded Recordings
  </h2>
  <ul
    class="publishedProfilesList"
  >
    <li
      class="publishedProfilesListItem"
    >
      <a
        class="publishedProfilesLink"
        href="/public/MACOSX/marker-chart/"
        title="Click here to load profile ⁨MacOS X profile⁩"
      >
        <div
          class="publishedProfilesDate"
        >
          Jul 5, 2020
        </div>
        <div
          class="publishedProfilesInfo"
        >
          <div
            class="publishedProfilesName"
          >
            <strong>
              MacOS X profile
            </strong>
             (
            38.0s
            )
          </div>
          <div
            class="profileMetaInfoSummary"
          >
            <div
              class="profileMetaInfoSummaryProductAndVersion"
              data-product="Firefox"
            >
              Firefox 62
            </div>
            <div
              class="profileMetaInfoSummaryPlatform"
              data-toolkit="macOS 10.12"
            >
              macOS 10.12
            </div>
          </div>
        </div>
      </a>
      <div
        class="publishedProfilesActionButtons"
      >
        <button
          class="publishedProfilesDeleteButton photon-button photon-button-default"
          disabled=""
          title="This profile cannot be deleted because we lack the authorization information."
          type="button"
        >
          Delete
        </button>
      </div>
    </li>
    <li
      class="publishedProfilesListItem"
    >
      <a
        class="publishedProfilesLink"
        href="/"
        title="Click here to load profile ⁨#012345⁩"
      >
        <div
          class="publishedProfilesDate"
        >
          2:00 PM
        </div>
        <div
          class="publishedProfilesInfo"
        >
          <div
            class="publishedProfilesName"
          >
            <strong>
              Profile #012345
            </strong>
             (
            2.0s
            )
          </div>
          <div
            class="profileMetaInfoSummary"
          >
            <div
              class="profileMetaInfoSummaryProductAndVersion"
              data-product="Fennec"
            >
              Fennec
            </div>
            <div
              class="profileMetaInfoSummaryPlatform"
              data-toolkit=""
            />
          </div>
        </div>
      </a>
      <div
        class="publishedProfilesActionButtons"
      >
        <div
          class="buttonWithPanel"
        >
          <button
            aria-expanded="false"
            class="buttonWithPanelButton publishedProfilesDeleteButton photon-button photon-button-default"
            title="Click here to delete the profile ⁨#012345⁩"
            type="button"
          >
            Delete
          </button>
        </div>
      </div>
    </li>
    <li
      class="publishedProfilesListItem"
    >
      <a
        class="publishedProfilesLink"
        href="/public/WINDOWS/marker-chart/"
        title="Click here to load profile ⁨Another good profile⁩"
      >
        <div
          class="publishedProfilesDate"
        >
          1:00 PM
        </div>
        <div
          class="publishedProfilesInfo"
        >
          <div
            class="publishedProfilesName"
          >
            <strong>
              Another good profile
            </strong>
             (
            38.0s
            )
          </div>
          <div
            class="profileMetaInfoSummary"
          >
            <div
              class="profileMetaInfoSummaryProductAndVersion"
              data-product="Firefox"
            >
              Firefox 77
            </div>
            <div
              class="profileMetaInfoSummaryPlatform"
              data-toolkit="Windows 10"
            >
              Windows 10
            </div>
          </div>
        </div>
      </a>
      <div
        class="publishedProfilesActionButtons"
      >
        <button
          class="publishedProfilesDeleteButton photon-button photon-button-default"
          disabled=""
          title="This profile cannot be deleted because we lack the authorization information."
          type="button"
        >
          Delete
        </button>
      </div>
    </li>
    <li
      class="publishedProfilesListItem"
    >
      <a
        class="publishedProfilesLink"
        href="/"
        title="Click here to load profile ⁨Layout profile⁩"
      >
        <div
          class="publishedProfilesDate"
        >
          Jul 3, 8:00 AM
        </div>
        <div
          class="publishedProfilesInfo"
        >
          <div
            class="publishedProfilesName"
          >
            <strong>
              Layout profile
            </strong>
             (
            0.0s
            )
          </div>
          <div
            class="profileMetaInfoSummary"
          >
            <div
              class="profileMetaInfoSummaryProductAndVersion"
              data-product="Firefox"
            >
              Firefox 68
            </div>
            <div
              class="profileMetaInfoSummaryPlatform"
              data-toolkit="Linux"
            >
              Linux
            </div>
          </div>
        </div>
      </a>
      <div
        class="publishedProfilesActionButtons"
      >
        <button
          class="publishedProfilesDeleteButton photon-button photon-button-default"
          disabled=""
          title="This profile cannot be deleted because we lack the authorization information."
          type="button"
        >
          Delete
        </button>
      </div>
    </li>
    <li
      class="publishedProfilesListItem"
    >
      <a
        class="publishedProfilesLink"
        href="/public/123abc456/calltree/?globalTrackOrder=0-1-2-3-4-5-6-7-8-9-10&hiddenGlobalTracks=1-2-3-4-5-6-7-8-9&localTrackOrderByPid=32027-1-2-0~12629-0~12149-0~11755-0~11945-0~12533-0~11798-0~11862-0~12083-0~12118-0~12298-0-1~&search=bla&thread=11&v=5"
        title="Click here to load profile ⁨#123abc⁩"
      >
        <div
          class="publishedProfilesDate"
        >
          May 20, 2018
        </div>
        <div
          class="publishedProfilesInfo"
        >
          <div
            class="publishedProfilesName"
          >
            <strong>
              Profile #123abc
            </strong>
             (
            0.0s
            )
          </div>
          <div
            class="profileMetaInfoSummary"
          >
            <div
              class="profileMetaInfoSummaryProductAndVersion"
              data-product="Firefox Preview"
            >
              Firefox Preview 80
            </div>
            <div
              class="profileMetaInfoSummaryPlatform"
              data-toolkit="Android 7"
            >
              Android 7
            </div>
          </div>
        </div>
      </a>
      <div
        class="publishedProfilesActionButtons"
      >
        <button
          class="publishedProfilesDeleteButton photon-button photon-button-default"
          disabled=""
          title="This profile cannot be deleted because we lack the authorization information."
          type="button"
        >
          Delete
        </button>
      </div>
    </li>
  </ul>
</main>
`;

exports[`UploadedRecordingsHome matches a snapshot when there is no published profiles 1`] = `
<main
  class="uploadedRecordingsHome"
>
  <header>
    <h1
      class="appHeader"
    >
      <span
        class="appHeaderSlogan"
      >
        <a
          class="appHeaderLink"
          href="/"
        >
          ⁨Firefox Profiler⁩
        </a>
         — 
        <span
          class="appHeaderSubtext"
        >
          Web app for ⁨Firefox⁩ performance analysis
        </span>
      </span>
      <a
        class="appHeaderGithubIcon"
        href="https://github.com/firefox-devtools/profiler"
        rel="noopener noreferrer"
        target="_blank"
        title="Go to our Git repository (this opens in a new window)"
      >
        <svg
          aria-label="github"
          class="octicon octicon-mark-github"
          height="22"
          version="1.1"
          viewBox="0 0 16 16"
          width="22"
        >
          <path
            d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"
            fill-rule="evenodd"
          />
        </svg>
      </a>
    </h1>
  </header>
  <h2
    class="photon-title-30"
  >
    Uploaded Recordings
  </h2>
  <p
    class="photon-body-30"
  >
    No profile has been uploaded yet!
  </p>
</main>
`;
